job_controller/status.go

```diff
@@ -0,0 +1,144 @@
+	// JobRunningReason is added in a job when it is running.
+	JobRunningReason = "JobRunning"
+	// JobFailedReason is added in a job when it is failed.
+	JobFailedReason = "JobFailed"
+	// JobRestarting is added in a job when it is restarting.
+	JobRestartingReason = "JobRestarting"
+
+	// labels for pods and servers.
+	ReplicaTypeLabel  = "replica-type"
+	ReplicaIndexLabel = "replica-index"
+	LabelGroupName    = "group-name"
+	LabelJobName      = "job-name"
+	LabelJobRole      = "job-role"
+)
+
+func isSucceeded(status common.JobStatus) bool {
+	return hasCondition(status, common.JobSucceeded)
+}
+
+func isFailed(status common.JobStatus) bool {
+	return hasCondition(status, common.JobFailed)
+}
+
+func hasCondition(status common.JobStatus, condType common.JobConditionType) bool {
+	for _, condition := range status.Conditions {
+		if condition.Type == condType && condition.Status == v1.ConditionTrue {
+			return true
+		}
+	}
+	return false
+}
+
+func updateJobConditions(jobStatus *common.JobStatus, conditionType common.JobConditionType, reason, message string) error {
+	condition := newCondition(conditionType, reason, message)
+	setCondition(jobStatus, condition)
+	return nil
+}
+
+// initializeReplicaStatuses initializes the ReplicaStatuses for replica.
+func initializeReplicaStatuses(jobStatus *common.JobStatus, rtype common.ReplicaType) {
+	if jobStatus.ReplicaStatuses == nil {
+		jobStatus.ReplicaStatuses = make(map[common.ReplicaType]*common.ReplicaStatus)
+	}
+
+	jobStatus.ReplicaStatuses[rtype] = &common.ReplicaStatus{}
+}
+
+// updateJobReplicaStatuses updates the JobReplicaStatuses according to the pod.
+func updateJobReplicaStatuses(jobStatus *common.JobStatus, rtype common.ReplicaType, pod *v1.Pod) {
+	switch pod.Status.Phase {
+	case v1.PodRunning:
+		jobStatus.ReplicaStatuses[rtype].Active++
+	case v1.PodSucceeded:
+		jobStatus.ReplicaStatuses[rtype].Succeeded++
+	case v1.PodFailed:
+		jobStatus.ReplicaStatuses[rtype].Failed++
+	}
+}
+
+// newCondition creates a new job condition.
+func newCondition(conditionType common.JobConditionType, reason, message string) common.JobCondition {
+	return common.JobCondition{
+		Type:               conditionType,
+		Status:             v1.ConditionTrue,
+		LastUpdateTime:     metav1.Now(),
+		LastTransitionTime: metav1.Now(),
+		Reason:             reason,
+		Message:            message,
+	}
+}
+
+// getCondition returns the condition with the provided type.
+func getCondition(status common.JobStatus, condType common.JobConditionType) *common.JobCondition {
+	for _, condition := range status.Conditions {
+		if condition.Type == condType {
+			return &condition
+		}
+	}
+	return nil
+}
+
+// setCondition updates the job to include the provided condition.
+// If the condition that we are about to add already exists
+// and has the same status and reason then we are not going to update.
+func setCondition(status *common.JobStatus, condition common.JobCondition) {
```

Comment:
Should have a unit test for this.

Reply:
This is called by updateJobConditions and already covered by TestUpdateJobConditions